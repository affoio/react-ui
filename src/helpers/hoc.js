import { createElement } from 'react';
import { omit } from 'ramda';

const componentFromProp = propName => {
  const Component = props => createElement(props[propName], omit([propName], props));
  Component.displayName = `componentFromProp(${propName})`;
  return Component;
};

export default componentFromProp;
