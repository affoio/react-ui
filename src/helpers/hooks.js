import { useEffect, useState } from 'react';

export const useOnClickOutside = (ref, handler) => {
  useEffect(() => {
    const listener = event => {
      // Do nothing if clicking ref's element or descendent elements
      if (!ref.current || ref.current.contains(event.target)) {
        return;
      }

      handler(event);
    };

    document.addEventListener('click', listener);
    document.addEventListener('touchstart', listener);
    document.addEventListener('focus', listener, true);

    return () => {
      document.removeEventListener('click', listener);
      document.removeEventListener('touchstart', listener);
      document.removeEventListener('focus', listener);
    };
  }, [ref, handler]);
};

export const useDebounce = (value, delay) => {
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [delay, value]);

  return debouncedValue;
};
