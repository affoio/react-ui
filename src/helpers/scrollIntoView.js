import smoothscroll from 'smoothscroll-polyfill';

smoothscroll.polyfill();

export function scrollIntoView(menuEl, focusedEl) {
  const menuRect = menuEl.getBoundingClientRect();
  const focusedRect = focusedEl.getBoundingClientRect();
  const overScroll = focusedEl.offsetHeight / 3;

  if (focusedRect.bottom + overScroll > menuRect.bottom) {
    menuEl.scrollTo(
      menuEl,
      Math.min(focusedEl.offsetTop + focusedEl.clientHeight - menuEl.offsetHeight + overScroll, menuEl.scrollHeight)
    );
  } else if (focusedRect.top - overScroll < menuRect.top) {
    menuEl.scrollTo(menuEl, Math.max(focusedEl.offsetTop - overScroll, 0));
  }
}
