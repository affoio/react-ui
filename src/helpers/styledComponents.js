import { css } from 'styled-components';

export const addStyleIfPropTruly = (propName, style) => props => {
  if (props[propName]) {
    return css`
      ${style}
    `;
  }
};
