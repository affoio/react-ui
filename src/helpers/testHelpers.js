import React from 'react';
import { ThemeProvider } from 'styled-components';
import { render } from 'enzyme';

import { MemoryRouter } from 'react-router';

export const renderWithThemeProvider = (children, theme = {}) =>
  render(<ThemeProvider theme={theme}>{children}</ThemeProvider>);

export const withProviders = (component, initialEntries, theme = {}) => (
  <MemoryRouter initialEntries={initialEntries}>
    <ThemeProvider theme={theme}>{component}</ThemeProvider>
  </MemoryRouter>
);
