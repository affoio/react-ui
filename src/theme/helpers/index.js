import { css } from 'styled-components';

export const isInteger = value => parseFloat(value) === parseInt(value, 10);
export const calculatePercentageGridValue = (value, gridSize) => `${(100 / gridSize) * value}%`;
export const getLimitedMedia = (breakpoints, current, currentIndex) => {
  return currentIndex !== Object.keys(breakpoints).length - 1
    ? `(min-width: ${breakpoints[current]}px) and (max-width: ${breakpoints[
        Object.keys(breakpoints)[currentIndex + 1]
      ] - 1}px)`
    : `(min-width: ${breakpoints[current]}px)`;
};

export const mediaQueryFunctionsList = breakpoints =>
  Object.keys(breakpoints).reduce((media, breakpoint, index) => {
    const newMedia = { ...media };
    newMedia[breakpoint] =
      index === 0
        ? (...args) => css`
            ${css(...args)}
          `
        : (...args) => css`
            @media (min-width: ${breakpoints[breakpoint]}px) {
              ${css(...args)}
            }
          `;
    return newMedia;
  }, {});

export const mediaQueryFunctionsListWithLimits = breakpoints =>
  Object.keys(breakpoints).reduce((media, breakpoint, index) => {
    const newMedia = { ...media };
    newMedia[breakpoint] = (...args) => css`
      @media ${getLimitedMedia(breakpoints, breakpoint, index)} {
        ${css(...args)}
      }
    `;
    return newMedia;
  }, {});
