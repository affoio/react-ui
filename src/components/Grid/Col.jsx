import { number, oneOfType, bool, node } from 'prop-types';
import styled, { css } from 'styled-components';

import { themeConfig } from '../../theme';
import { mediaQueryFunctionsListWithLimits, isInteger, calculatePercentageGridValue } from '../../theme/helpers';
import { getBreakpoints, getGridSize, getBaseIndent, getGutterWidth } from '../../theme/selectors';

const ColPropType = oneOfType([number, bool]);
const ColOffsetPropType = number;

const colProps = themeConfig.breakpointNames.reduce((colProp, breakpoint) => {
  const newColProp = { ...colProp };
  newColProp[breakpoint] = ColPropType;
  newColProp[`${breakpoint}Offset`] = ColOffsetPropType;

  return newColProp;
}, {});

const propTypes = {
  ...colProps,
  children: node,
};

const createBreakpointsStyles = props => {
  const { theme } = props;
  const gridSize = getGridSize(theme);
  const breakpoints = getBreakpoints(theme);
  const mediaQueries = mediaQueryFunctionsListWithLimits(breakpoints);

  return Object.keys(props)
    .filter(breakpoint => themeConfig.breakpointNames.indexOf(breakpoint) > -1)
    .sort(
      (breakpoint1, breakpoint2) =>
        themeConfig.breakpointNames.indexOf(breakpoint1) - themeConfig.breakpointNames.indexOf(breakpoint2)
    )
    .map(breakpoint => {
      return isInteger(props[breakpoint])
        ? mediaQueries[breakpoint]`
            flex-basis: ${calculatePercentageGridValue(props[breakpoint], gridSize)};
            max-width: ${calculatePercentageGridValue(props[breakpoint], gridSize)};
          `
        : mediaQueries[breakpoint]`${
            props[breakpoint] === true
              ? `
            flex-grow: 1;
            flex-basis: 0;
            max-width: 100%;
            `
              : `
              display: none;
            `
          }`;
    });
};

const createBreakpointOffsetStyles = props => {
  const { theme } = props;

  return Object.keys(props)
    .filter(breakpoint => themeConfig.breakpointOffsetNames.indexOf(breakpoint) > -1)
    .map(breakpoint => {
      return mediaQueryFunctionsListWithLimits(getBreakpoints(theme))[breakpoint.replace(/Offset/g, '')]`
        margin-left: ${calculatePercentageGridValue(props[breakpoint], getGridSize(theme))};
      `;
    });
};

const Col = styled.div`
  box-sizing: border-box;
  ${({ grow }) => `flex-grow: ${+!!grow};`}
  max-width: 100%;

  ${({ theme }) => {
    const colPadding = `${getGutterWidth(theme) * getBaseIndent(theme)}px`;
    return css`
      padding-right: ${colPadding};
      padding-left: ${colPadding};
    `;
  }}
  flex-direction: column;
  ${createBreakpointsStyles}
  ${createBreakpointOffsetStyles}
`;

Col.propTypes = propTypes;

export default Col;
