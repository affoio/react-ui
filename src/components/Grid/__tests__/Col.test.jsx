import React from 'react';
import Col from '../Col';
import { renderWithThemeProvider } from '../../../helpers/testHelpers';

describe('<Col>', () => {
  it('renders Col', () => {
    const tree = renderWithThemeProvider(<Col>Test Text</Col>);

    expect(tree).toMatchSnapshot();
  });

  it('renders Col with grow', () => {
    const tree = renderWithThemeProvider(<Col grow>Test Text</Col>);

    expect(tree).toMatchSnapshot();
  });

  it('renders hidden in xs Col', () => {
    const tree = renderWithThemeProvider(<Col xs={false}>Test Text</Col>);

    expect(tree).toMatchSnapshot();
  });

  it('renders Col fullsize in xs', () => {
    const tree = renderWithThemeProvider(<Col xs>Test Text</Col>);

    expect(tree).toMatchSnapshot();
  });

  it('renders Col xs=12 sm=6 lg=4', () => {
    const tree = renderWithThemeProvider(
      <Col xs={12} sm={6} lg={4}>
        Test Text
      </Col>
    );

    expect(tree).toMatchSnapshot();
  });
  it('renders Col withOffset xs=12 sm=6 lg=4', () => {
    const tree = renderWithThemeProvider(
      <Col xsOffset={12} smOffset={6} lgOffset={4}>
        Test Text
      </Col>
    );

    expect(tree).toMatchSnapshot();
  });
});
