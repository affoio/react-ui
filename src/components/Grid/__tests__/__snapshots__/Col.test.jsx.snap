// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Col> renders Col 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 100%;
  padding-right: 8px;
  padding-left: 8px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  class="c0"
>
  Test Text
</div>
`;

exports[`<Col> renders Col fullsize in xs 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 100%;
  padding-right: 8px;
  padding-left: 8px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (min-width:320px) and (max-width:767px) {
  .c0 {
    -webkit-box-flex: 1;
    -webkit-flex-grow: 1;
    -ms-flex-positive: 1;
    flex-grow: 1;
    -webkit-flex-basis: 0;
    -ms-flex-preferred-size: 0;
    flex-basis: 0;
    max-width: 100%;
  }
}

<div
  class="c0"
>
  Test Text
</div>
`;

exports[`<Col> renders Col with grow 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  max-width: 100%;
  padding-right: 8px;
  padding-left: 8px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  class="c0"
>
  Test Text
</div>
`;

exports[`<Col> renders Col withOffset xs=12 sm=6 lg=4 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 100%;
  padding-right: 8px;
  padding-left: 8px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (min-width:320px) and (max-width:767px) {
  .c0 {
    margin-left: 100%;
  }
}

@media (min-width:768px) and (max-width:1279px) {
  .c0 {
    margin-left: 50%;
  }
}

@media (min-width:1280px) {
  .c0 {
    margin-left: 33.333333333333336%;
  }
}

<div
  class="c0"
>
  Test Text
</div>
`;

exports[`<Col> renders Col xs=12 sm=6 lg=4 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 100%;
  padding-right: 8px;
  padding-left: 8px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (min-width:320px) and (max-width:767px) {
  .c0 {
    -webkit-flex-basis: 100%;
    -ms-flex-preferred-size: 100%;
    flex-basis: 100%;
    max-width: 100%;
  }
}

@media (min-width:768px) and (max-width:1279px) {
  .c0 {
    -webkit-flex-basis: 50%;
    -ms-flex-preferred-size: 50%;
    flex-basis: 50%;
    max-width: 50%;
  }
}

@media (min-width:1280px) {
  .c0 {
    -webkit-flex-basis: 33.333333333333336%;
    -ms-flex-preferred-size: 33.333333333333336%;
    flex-basis: 33.333333333333336%;
    max-width: 33.333333333333336%;
  }
}

<div
  class="c0"
>
  Test Text
</div>
`;

exports[`<Col> renders hidden in xs Col 1`] = `
.c0 {
  box-sizing: border-box;
  -webkit-box-flex: 0;
  -webkit-flex-grow: 0;
  -ms-flex-positive: 0;
  flex-grow: 0;
  max-width: 100%;
  padding-right: 8px;
  padding-left: 8px;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

@media (min-width:320px) and (max-width:767px) {
  .c0 {
    display: none;
  }
}

<div
  class="c0"
>
  Test Text
</div>
`;
